segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkIterator.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import io.pravega.common.concurrent.Futures;
+import io.pravega.segmentstore.storage.metadata.ChunkMetadata;
+import io.pravega.segmentstore.storage.metadata.MetadataTransaction;
+import io.pravega.segmentstore.storage.metadata.SegmentMetadata;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+/**
+ * Helper class for iterating over list of chunks.
+ */
+class ChunkIterator {
+    private final ChunkedSegmentStorage chunkedSegmentStorage;
+    private final MetadataTransaction txn;
+    private volatile String currentChunkName;
+    private volatile ChunkMetadata currentMetadata;
+
+    ChunkIterator(ChunkedSegmentStorage chunkedSegmentStorage, MetadataTransaction txn, SegmentMetadata segmentMetadata) {
+        this.chunkedSegmentStorage = chunkedSegmentStorage;
+        this.txn = txn;
+        currentChunkName = segmentMetadata.getFirstChunk();
+    }
+
+    public CompletableFuture<Void> forEach(BiConsumer<ChunkMetadata, String> consumer) {
+        return Futures.loop(
+                () -> currentChunkName != null,
+                () -> txn.get(currentChunkName)
+                        .thenApplyAsync(storageMetadata -> {
+                            currentMetadata = (ChunkMetadata) storageMetadata;
+                            consumer.accept(currentMetadata, currentChunkName);
+                            // Move next
+                            currentChunkName = currentMetadata.getNextChunk();
+                            return null;
+                        }, chunkedSegmentStorage.getExecutor()),
```

Comment:
I think The reason for not using non-async overload is that they *can* end up running on some Fork Join pool or caller pool.
By providing the same executor we are sure it will only run on given thread pool, which is what we want.